vllm/v1/sample/ops/topk_topp_sampler.py

```diff
@@ -122,23 +126,65 @@ def forward_tpu(
         k: Optional[torch.Tensor],
         p: Optional[torch.Tensor],
     ) -> torch.Tensor:
-        # If only top-k is specified, use pytorch's builtin topk op. This leads
-        # to significant speed up on TPU compared to using apply_top_k_top_p.
-        if k is not None and p is None:
-            topk_values, topk_indices = torch.topk(logits, k, dim=-1)
-
-            mask = torch.ones_like(logits, dtype=torch.bool)
-            mask.scatter_(-1, topk_indices, False)
-            logits.masked_fill_(mask, float('-inf'))
-        else:
-            # TODO Placeholder for TPU optimized topp kernel
-            # logits = apply_top_k_top_p(logits, k, p)
-            pass
-
+        logits = apply_top_k_top_p_tpu(logits, k, p)
         probs = logits.softmax(dim=-1, dtype=torch.float32)
         return random_sample(probs, generators)
 
 
+def apply_top_k_top_p_tpu(
+    logits: torch.Tensor,
+    k: torch.Tensor,
+    p: torch.Tensor,
+) -> torch.Tensor:
+    if k is not None:
+        logits = apply_top_k_only(logits, k)
+
+    if p is not None:
+        logits = apply_approx_top_p(logits, p)
```

Comment:
is this top-p (to be used as topp-only) implementation also needed on gpu? @njhill

Reply:
IIUC scattering isn't a bottleneck on GPU so this impl wouldn't bring much benefit (plus, this impl still involves a full vocab sort same as the forward_native version).